src/Microsoft.AspNet.OData.Shared/Query/DefaultSkipTokenHandler.cs

```diff
@@ -0,0 +1,387 @@
+                    {
+                        return GenerateSkipTokenValue(obj, model, orderByNodes);
+                    };
+                    
+                    return GetNextPageHelper.GetNextPageLink(baseUri, pageSize, instance, skipTokenGenerator);
+                }
+
+                if (context.QueryOptions != null && context.QueryOptions.OrderBy != null)
+                {
+                    orderByNodes = context.QueryOptions.OrderBy.OrderByNodes;
+                }
+
+                skipTokenGenerator = (obj) =>
+                {
+                    return GenerateSkipTokenValue(obj, model, orderByNodes);
+                };
+            }
+            return context.InternalRequest.GetNextPageLink(pageSize, instance, skipTokenGenerator);
+        }
+
+        /// <summary>
+        /// Returns a function that converts an object to a skiptoken value string
+        /// </summary>
+        /// <param name="lastMember"> Object based on which SkipToken value will be generated.</param>
+        /// <param name="model">The edm model.</param>
+        /// <param name="orderByNodes">QueryOption </param>
+        /// <returns></returns>
+        protected virtual string GenerateSkipTokenValue(Object lastMember, IEdmModel model, IList<OrderByNode> orderByNodes)
+        {
+            object value;
+            if (lastMember == null)
+            {
+                return String.Empty;
+            }
+            IEnumerable<IEdmProperty> propertiesForSkipToken = GetPropertiesForSkipToken(lastMember, model, orderByNodes);
+
+            String skipTokenvalue = String.Empty;
+            if (propertiesForSkipToken == null)
+            {
+                return skipTokenvalue;
+            }
+
+            int count = 0;
+            int lastIndex = propertiesForSkipToken.Count() - 1;
+            foreach (IEdmProperty property in propertiesForSkipToken)
+            {
+                bool islast = count == lastIndex;
+                string propertyName = property.Name;
+                IEdmStructuredObject obj = lastMember as IEdmStructuredObject;
+                if (obj != null)
+                {
+                    obj.TryGetPropertyValue(property.Name, out value);
+                }
+                else
+                {
+                    value = lastMember.GetType().GetProperty(property.Name).GetValue(lastMember);
+                }
+
+                String uriLiteral = String.Empty;
+                if (value == null)
+                {
+                    uriLiteral = ODataUriUtils.ConvertToUriLiteral(value, ODataVersion.V401);
+                }
+                else if (TypeHelper.IsEnum(value.GetType()))
+                {
+                    ODataEnumValue enumValue = new ODataEnumValue(value.ToString(), value.GetType().FullName);
+                    uriLiteral = ODataUriUtils.ConvertToUriLiteral(enumValue, ODataVersion.V401, model);
+                    uriLiteral = "'enumType'" + uriLiteral;
```

Comment:
>uriLiteral = "'enumType'" + uriLiteral; [](start = 20, length = 39)

I think it's enough to use the string from "ConvertToUriLiterla", prefix "enumType" is useless? So, what's the reason to do like this?

Reply:
As mentioned above. Getting the type reference is expensive. We don't want to do it for all types since type reference is only needed for enums. 

---
In reply to: [282594292](https://github.com/OData/WebApi/pull/1724#discussion_r282594292) [](ancestors = 282594292)